docs/02-app/01-building-your-application/07-configuring/03-environment-variables.mdx

```diff
@@ -135,6 +135,32 @@ const env = process.env
 setupAnalyticsService(env.NEXT_PUBLIC_ANALYTICS_ID)
 ```
 
+### Runtime Environment Variables
+
+Next.js can support both build time and runtime environment variables.
+
+**By default, environment variables are only available on the server**. To expose an environment variable to the browser, it must be prefixed with `NEXT_PUBLIC_`. However, these public environment variables will be inlined into the JavaScript bundle during `next build`.
+
+To read runtime environment variables, we recommend using `getServerSideProps` or [incrementally adopting the App Router](/docs/app/building-your-application/upgrading/app-router-migration). With the App Router, we can safely read environment variables on the server during dynamic rendering. This allows you to use a singular Docker image that can be promoted through multiple environments with different values.
+
+```jsx
+import { unstable_noStore as noStore } from 'next/cache';
+
+export default function Component() {
+  noStore();
+  // cookies(), headers(), and other dynamic functions
+  // will also opt into dynamic rendering, making
+  // this env variable is evaluated at runtime
+  const value = process.env.MY_VALUE
+  // ...
```

Comment:
Hey @leerob 

This is a really welcome change, I've been really looking forward to have something like this. One question here, does this approach cover the use case of using the env vars within client components? To elaborate, we have had cases where our entire page is a client component, and within the page we are using client-facing env variables. Since Next.js bundles the `NEXT_PUBLIC` variables at build time, there is no way for us to use browser-facing variables on the client.

One example of using this - we have a Playwright test suite - our thinking was to do a `next build`, then locally start that build, perform e2e tests on that build output (I know that Playwright has a dev server option, we strictly want to execute tests on the build output to mimic the production environment as best as possible), and then after the tests are done, just convert that build into a docker image and deploy to production. With the current approach, this requires to create a new Next.js build output every time the environment variables change, which means in this case we have to create two builds instead of reusing the one.

It might be worth including a rudimentary example documenting how this would be solved. For example, Remix docs simply suggest to use the loaders to fetch it from the server and then expose it to the browser, pretty much like we had with `getServerSideProps` in the Pages Router. It would be really beneficial to explain how to do this with the App Router. Is the solution to just wrap that piece of client-side code in a server component and perform the fetching in that server component? What approach do you recommend in this case?

Reply:
> Is the solution to just wrap that piece of client-side code in a server component and perform the fetching in that server component?

I think this is exactly what is documented in this PR (link: https://github.com/vercel/next.js/pull/58027/files#diff-def283aebd0ab1897890d36b399b369fd6cbdb8aa384e8a85de7d67a470d5473R101)

Reply:
> > Is the solution to just wrap that piece of client-side code in a server component and perform the fetching in that server component?
> 
> I think this is exactly what is documented in this PR (link: [#58027 (files)](https://github.com/vercel/next.js/pull/58027/files#diff-def283aebd0ab1897890d36b399b369fd6cbdb8aa384e8a85de7d67a470d5473R101))

Yes, that's why I asked whether the `noStore` approach is applicable for client components.

Reply:
I am pretty sure it can't be used inside of client components.

Reply:
Yeah, I just tried it within a client component and it doesn't seem to have any effect.

Reply:
I believe what you're looking for here would be a client component that consumes a `NEXT_PUBLIC_` env var. Correct?